src/lib/party.ts

```diff
@@ -692,14 +692,17 @@ class Party {
     options?: { isSameAsFeePayer?: Bool | boolean; nonce?: UInt32 }
   ) {
     let { nonce, isSameAsFeePayer } = options ?? {};
-    // if not specified, optimistically determine isSameAsFeePayer from the current transaction
-    // (gotcha: this makes the circuit depend on the fee payer parameter in the transaction.
-    // to avoid that, provide the argument explicitly)
+    // if not specified, determine isSameAsFeePayer from the current transaction
+    // (gotcha: this doesn't constrain `isSameAsFeePayer`, to avoid making the circuit depend on something that can't be
+    // inferred at compile time. to constrain it, provide the argument explicitly)
```

Comment:
what does constraining it look like? What are the implications to leaving it unconstrained?

Reply:
Is there something that developers need to understand here to write safe contracts?

Reply:
`isSameAsFeePayer` determines if we increment the nonce precondition by one, because the fee payer already increments the nonce once. For some reason I thought that we should do that inside the circuit. Now that I think about it more clearly, I see no reason at all to do the extra nonce incrementing in the circuit. There's no attack vector to modifying the nonce -- we set a nonce precondition anyway, so it has to be the current one to be accepted. I think this comment is just me being confused and erring on the paranoid side when writing code that will end up in smart contract circuits :D

So I'm inclined to delete those comments, move the nonce computation into a `Circuit.witness` block entirely and even remove the `isSameAsFeePayer` argument.

It gets more complicated though, because incrementing the nonce here is the wrong behavior now, to be addressed in https://github.com/o1-labs/snarkyjs/issues/271. In the future, `isSameAsFeePayer` will determine whether to set `useFullCommitment` to `true`, or do a nonce increment. So there are _four_  variables which would all get determined at runtime from whether a fee payer is set in snarkyjs:
* party.body.useFullCommitment (a `Bool`)
* party.body.incrementNonce (a `Bool`)
* party.body.preconditions.account.nonce.lower
* party.body.preconditions.account.nonce.upper

Maybe the same reasoning might apply in the future as well. We can let the prover set all four variables to whatever they want. By being less restrictive -- setting both useFullCommitment and incrementNonce to false -- they can make their transaction replay-able. However, they can only create this transaction if they are owning the signing party's private key, and the signing party is the only account that is potentially harmed by replaying this transaction. So, again I see no attack vector.

Reply:
TL;DR, I think we can avoid leaking these complications to developers and still have them write perfectly safe contracts